[WIP] add kor jpn chn font for rgui

trngaje opened the pull request:
## Description

created Korean, Japanese, and Chinese fonts for rgui drivers.The font size is 10 pixels * 10 pixels.
tested in rpi3,4 and oga.Corrected static analysis errors as well.
Previous PR is as follows
- https://github.com/libretro/RetroArch/pull/11589

Added definition HAVE_CKJ_BITMAP_FONTS
Transferring code implemented in blit_line_regular() to blit_line_extended()
Font bin files must exist in the path below
~/.config/retroarch/assets/rgui/font/
rgui_font_kor10x10.bin
rgui_font_jpn10x10.bin
rgui_font_chn10x10.bin

## Related Pull Requests

libretro/retroarch-assets#389

## Reviewers

@jdgleaver

jdgleaver commented:
@trngaje Thank you for this! From a first glance, it is a *huge* improvement. I shall review this properly as soon as possible, a little later in the week.

jdgleaver commented:
@trngaje Okay, there are a few smaller issues that should be quite easy to deal with (I can probably fix them myself, and give you a patch to apply to the PR) - we can come to these at a later date.

But there's one problem that's larger, and needs some more thought:

I was testing the PR, and found that your changes to `gfx_animation.c` and `encoding_utf.c` actually break all ticker text when not using RGUI - *and* they don't really improve ticker text in RGUI either (lengths are still wrong, and the scrolling stutters).

It took a little while to identify the issue, but it seems to be this: you're effectively trying to plug a variable width font into RGUI, but RGUI is not designed for this, and cannot support it without a major redesign - which I don't think we want to do.

I think the correct (and certainly the easiest) approach is to instead use internal  `font_width`, `font_width_stride`, `font_height` and `font_height_stride` variables in RGUI, which are set to the *maximum* values for the currently selected language - so that's `width x height =  10x10` for CKJ languages, and `5x10` for everything else. This means the entire 'grid size' of RGUI will increase when using your fonts, and everything remains fixed width, and the whole menu should just work as normal - with no complications.

Another benefit of this is that all your changes to `gfx_animation.c` and `encoding_utf.c` will no longer be required, and can be removed - which means your PR will no longer have any real negative performance impact unless the user is explicitly enabling extended ASCII in RGUI (this is very desirable!)

The main downside is that, to prevent non-translated text from looking 'wonky', we're going to need a new 10x10 ASCII/Extended-ASCII bitmap font which will be loaded in place of the regular one whenever a CKJ language is selected - i.e. RGUI's `font_lut` should be allocated dynamically, and in `rgui_init_font_lut()` we populate it as normal if a non-CKJ language is used, but copy from a new 10x10 'bitmap_bin' otherwise. Do you see what I mean?

So the question is: would you be able to find and add this new non-CKJ btimap font, and wire it up in RGUI? If you can, I think we will be in excellent shape :)

trngaje commented:
> So the question is: would you be able to find and add this new non-CKJ btimap font, and wire it up in RGUI? If you can, I think we will be in excellent shape :)

I am not able to review this issue at all because I work overtime today.
Do you want to apply ascii & latin extended font as 10x10?
Do you want to load 5x10 fonts in English and 10x10 dynamically in CJK fonts?
I think it will be possible to apply 10x10 ascii because most of the Korean font contains English font.
However, I think I will know how it would look if I test it once because the width of the asian letter is usually twice as wide as English.
I tried to apply Russian font in size 5x10 but it wasn't pretty.
All the letters except Arabic and Persian can be converted as long as there is a ttf font designed in a small size. Arabic or Persian is RTL, so it's not easy to implement.

jdgleaver commented:
> I am not able to review this issue at all because I work overtime today.

Certainly, there is no hurry to do anything, and please don't feel pressured by my requests - contributing to RA should be fun, so only work on this when you have time, and when it is convenient. We are happy to wait however long it takes :)

> Do you want to apply ascii & latin extended font as 10x10?
> Do you want to load 5x10 fonts in English and 10x10 dynamically in CJK fonts?

Yes, this is correct:

- When a non-CJK language is selected, we use the normal existing 5x10 font for ASCII & extended Latin

- When a CJK language is selected, we instead use a new 10x10 font for ASCII & extended Latin - so we basically need a 10x10 version of this: https://raw.githubusercontent.com/libretro/RetroArch/master/gfx/drivers_font_renderer/bitmap.bmp

By doing this, we ensure that characters rendered in RGUI always have the same dimensions (either 5x5 or 10x10), which prevents anything from breaking.

> I tried to apply Russian font in size 5x10 but it wasn't pretty.
> All the letters except Arabic and Persian can be converted as long as there is a ttf font designed in a small size. Arabic or Persian is RTL, so it's not easy to implement.

This is fine - CJK fonts are more than enough for this PR. If you want to try making other 10x10 fonts for other languages (for Russian, perhaps) at some point in the future, that could be nice - but sure, let's not worry about Arabic or Persian if it's impractical. :)

trngaje commented:
> Certainly, there is no hurry to do anything, and please don't feel pressured by my requests - contributing to RA should be fun, so only work on this when you have time, and when it is convenient. We are happy to wait however long it takes :)

@jdgleaver
It's fun to solve a problem that no one touches. And I am modifying myself because I like it. The most difficult part of this work was using git. It's not a long task, so you don't have to worry.

trngaje commented:
[rgui_font_eng10x10.zip](https://github.com/libretro/RetroArch/files/5612069/rgui_font_eng10x10.zip)

trngaje commented:
@jdgleaver 
10x10 English font applied. And removed unused codes.

jdgleaver commented:
Thanks for the update. This is much better now that it's confined to a limited set of files.

Unfortunately there are still some fundamental errors in the way that it's implemented in RGUI (I think you misunderstood what I said about increasing the 'grid size' of RGUI - it needs to be done 'globally', not just in the line renderers and ticker). If you run your code, the menu is immediately and very obviously broken, from the very first screen... (look at the clock, for example, and the misaligned title, and the incorrect width of the 'value' text)

I think it might be best if I take the good work you've done here and fix the errors myself. But before I do this, would it be at all possible to improve the 10x10 English font?

![Screenshot_2020-11-30_12-32-53](https://user-images.githubusercontent.com/38211560/100612616-2d252d00-330b-11eb-81a0-430914b51cf7.png)

As shown in the above screenshot, it's not quite ready for a production release - many of the characters have strong aliasing artefacts (`N`, `M` and anything with diagonal lines are particularly affected), `e` looks like theta and hyphens are shown as underscores.

trngaje commented:
> would it be at all possible to improve the 10x10 English font?

@jdgleaver 
The font I shared here is part of what I worked on.
If I match the alignment of a letter, some characters cannot be expressed within a set size.
(ex, p q g ...)
I will convert rgui original English font into a font binary file, including a space of 10x10.
Perhaps a lot of empty spaces are marked, but the basic font expresses English best without aliasing.

> (I think you misunderstood what I said about increasing the 'grid size' of RGUI - it needs to be done 'globally', not just in the line renderers and ticker).
I don't understand it properly. Because I am a beginner about retroarch. And I don't have basic knowledge of fonts. It's a step to learn one by one.

Recently, rg351p device has arrived and is spending a lot of time setting it up.

Take your time. I expect a offical release next year. I thought it wouldn't be easy, and I'm happy to have someone to review it with me.

trngaje commented:
[rgui_font_eng10x10.zip](https://github.com/libretro/RetroArch/files/5617035/rgui_font_eng10x10.zip)

@jdgleaver  
10x10 binary generated from the original font.

jdgleaver commented:
> If I match the alignment of a letter, some characters cannot be expressed within a set size.
(ex, p q g ...)

I think the letter alignment is okay, actually - it's only the 'jaggy' characters that look rough (and the double-thick vertical lines on N, M). I know that editing fonts is fiddly and annoying, so fixing this new font may not be easy... If you prefer to pad the original RGUI font, that's fine with me. It's entirely up to you, really - my only concern is that looks 'clean', like the rest of RetroArch :)

> I don't understand it properly. Because I am a beginner about retroarch. And I don't have basic knowledge of fonts. It's a step to learn one by one.

This is absolutely fine - and I want to make it clear that I wasn't criticising you! It is definitely the case that the errors here are only due to inexperience with RetroArch (not lack of coding skill!), and that's why I want to fix them myself - because I don't think it's fair for a new contributor to have to deal with so many obtuse RetroArch-specific issues in their first PR.

All the groundwork that you have done here is excellent, and we would never have had CJK fonts in RGUI without your efforts!

> Take your time. I expect a offical release next year. I thought it wouldn't be easy, and I'm happy to have someone to review it with me.

Thank you for your patience. We will get this finished in due course, and I think many people will be happy with the end result.

> Recently, rg351p device has arrived and is spending a lot of time setting it up.

Ah! I hope you enjoy your new device :)

jdgleaver commented:
Ah, good job! This new font looks fine :)

jdgleaver commented:
@trngaje I'm currently refactoring this PR. It needed quite a few changes, but it's going well - I still have some optimisation work to do, but I hope to have it finished tomorrow. Overall I am very pleased with the work you did here :)

One thing I noticed, however: the bitmap fonts have limited Unicode coverage. I think the Korean one is fine, but the Japanese font seems to be missing mostly punctuation characters (I guess this isn't vitally important?) - and the Chinese font has many blanks, such that the menu isn't very usable in this language.

Do you think it would be possible to add some of these missing characters, particularly for the Chinese font? I think this could be done in stages, after the code itself is merged. Or is there insufficient source material to work from? (I imagine a complete Chinese 10x10 bitmap font is quite a body of work...)

trngaje commented:
> Do you think it would be possible to add some of these missing characters, particularly for the Chinese font?
Yes, it's possible.
Chinese uses several code pages, but I have implemented only the letters that Japanese and Chinese use for common purposes.

> but the Japanese font seems to be missing mostly punctuation characters
No Japanese punctuation related code page font generated. I just checked the possibility.


At first, I started working with 8x8 Korean font. The result is very good because the size of Korean and English are similar. But I couldn't get a font developed in small size in Chinese, Japanese, so I grew the font size to 10x10.
Chinese characters used in Japanese are often tuned for games, but Chinese characters in Chinese are raw. It's not pretty even if you display it. So, I implemented only the code areas that are used a lot as a preview. 

I can create a font binary if you want to be inserted all the Chinese codes. If you want, I can add all the Japanese punctuation marks included in the font.

Everything's going to start after work, but...

> I'm currently refactoring this PR. It needed quite a few changes, but it's going well - I still have some optimisation work to do, but I hope to have it finished tomorrow.
That's great news.

> Overall I am very pleased with the work you did here
I'm so glad you're interested in my work.

jdgleaver commented:
Thanks for the info. I don't know any Japanese/Chinese, so it's great to have someone with your knowledge around. I guess the 'common' glyphs should have been enough, but the translators who worked on the Japanese and Chinese text seem to have used many uncommon characters!

I think as complete a set as possible of the Japanese characters in the range 0x3040-0x30FF, and Chinese characters in the range 0x4E00-0x9FFF would be good, to minimise gaps in the menu (so the translators can use glyphs freely). But again, this can be done in stages - there's no hurry, and you can do it however you feel is easiest (all at once, or in batches, or just by looking at the menu itself to selective add obviously missing characters). And of course, if adding certain glyphs is impractical, we can leave them out.

The code will all be in place - provided the font dimensions and unicode ranges stay the same, it's just a matter of uploading new bins to retroarch-assets as they become available.

trngaje commented:
@jdgleaver 
There are characters in use in the menu that do not contain font content.
It is as follows.
“啟” (U+555F)
“檔” (U+6A94)
There may be more because you didn't search all the menu statements.
I'm not sure because I don't have many small Chinese fonts, but I'll google them and fill them up.

jdgleaver commented:
@trngaje I've just made the initial PRs for the refactored CJK font support:

https://github.com/libretro/RetroArch/pull/11677
https://github.com/libretro/retroarch-assets/pull/393

I would appreciate it if you could test the code :)

> There are characters in use in the menu that do not contain font content.
> It is as follows.
> “啟” (U+555F)
> “檔” (U+6A94)
> There may be more because I didn't search all the menu statements.
> I'm not sure because I don't have many small Chinese fonts, but I'll google them and fill them up.

Thank you for finding these, and for offering to add bitmaps. Yes, it is difficult to search all the text in the menu - I think we may need to just fill in the blanks piecemeal as they become apparent...

> What is the truetype Chinese font used by xmb?
> I think it would be best to extract the letter from the font.

The default XMB font can be found here: https://github.com/libretro/retroarch-assets/tree/master/xmb/monochrome

trngaje commented:
@jdgleaver 
The filenames are wrong in assets after update assets

mv rgui_font_chn10x10.bin bitmap10x10_chn.bin
mv rgui_font_eng10x10.bin bitmap10x10_eng.bin
mv rgui_font_kor10x10.bin bitmap10x10_kor.bin
mv rgui_font_jpn10x10.bin bitmap10x10_jpn.bin

jdgleaver commented:
@trngaje Yes, I changed the names in the assets PR (https://github.com/libretro/retroarch-assets/pull/393), because the new 10x10 bitmap code is 'generic', and could be used in other places too (e.g. for OSD text, not just RGUI). I therefore wanted to remove the `rgui_` prefix from the bin files themselves.

The new names are referenced here: https://github.com/libretro/RetroArch/blob/master/gfx/drivers_font_renderer/bitmapfont_10x10.c

If you update assets from the online updater, you should get the current (correctly named) files.

Note that I also compressed them, to save a bit of space :)

If you add more characters to these files, we should use the new names. I hope this is okay...?

trngaje commented:
@jdgleaver 
> If you add more characters to these files, we should use the new names. I hope this is okay...?

I don't care what name is.

I was mistaken that the changed font would have been downloaded after update assets.
Can I read both uncompressed fonts and compressed fonts in code?
I already checked the modified code through the diff.
Thank you for adding my nickname to Copyright.
It seems like all my modifications have been lost by the refactoring but I am proud of myself.

jdgleaver commented:
@trngaje 

> Can I read both uncompressed fonts and compressed fonts in code?

Yes, absolutely - the rzip reader handles both, so you can used uncompressed files while testing any character additions

> Thank you for adding my nickname to Copyright.

It was the very least I could do!

> It seems like all my modifications have been lost by the refactoring but I am proud of myself.

You should be very proud of yourself.

RetroArch is 10 years old, and in all that time we have never had CJK font support in RGUI. That means RetroArch on all platforms without hardware renderers has been completely inaccessible for Chinese/Japanese/Korean speakers - that's potentially millions of users who couldn't enjoy the application.

Now that issue is fixed - and it's solely because of you. No one else had thought of doing what you did. I could never have done it without using your code as a template. You have made an absolutely huge contribution, and so many people will be thankful for it.

And your modifications haven't been lost - I only moved them around!

The bulk of my changes are all RA integration things, and performance tweaks - things that would have been entirely unfair for you to have to deal with in your first PR. These are things that you can slowly become familiar with, as and when you make other PRs.

So once again, the project cannot thank you enough for your work. And I so very much look forward to your additions to the font files. Let me know whenever you make another assets PR, and I will ensure it gets merged promptly :)

trngaje commented:
@jdgleaver 
[bitmap10x10_chn.zip](https://github.com/libretro/RetroArch/files/5682927/bitmap10x10_chn.zip)

jdgleaver commented:
This new font is fantastic! Thank you very much. We should definitely add this to retroarch-assets - would you like to PR it, or should I?

I think all we're missing now is this punctuation block: https://www.compart.com/en/unicode/block/U+3000

This is a bit tricky, since it's separated from the exiting CJK font ranges. I believe, however, that we could add it to the beginning of the Japanese font, changing `FONT_10X10_SIZE_JPN` and `FONT_10X10_GLYPH_MIN_JPN` as appropriate. Would you be interested in doing this? Absolutely no hurry, of course - you've already done so much, so please take a break if you want to :)

It's very exciting to have so much Unicode coverage in RGUI.

trngaje commented:
> This new font is fantastic! Thank you very much. We should definitely add this to retroarch-assets - would you like to PR it, or should I?

Do it yourself. because I'm not used to git command.

> I think all we're missing now is this punctuation block: https://www.compart.com/en/unicode/block/U+3000

Can I make it into a separate font? What should I name the file?



> Would you be interested in doing this? Absolutely no hurry, of course - you've already done so much, so please take a break if you want to

Don't worry, it's for the purpose of resting. 

I'll make the font easily, so please apply it right away. It is difficult for me to make the new BR after deleting the existing branch and making the master up to date.

trngaje commented:
@jdgleaver 

The punctuation mark is partially contained in Chinese and Japanese fonts.
I will create it manually and attach it to the Japanese font.

trngaje commented:
@jdgleaver 
[bitmap10x10_jpn.zip](https://github.com/libretro/RetroArch/files/5683261/bitmap10x10_jpn.zip)

#define FONT_10X10_SIZE_JPN      3328
#define FONT_10X10_GLYPH_MIN_JPN 0x3000
#define FONT_10X10_GLYPH_MAX_JPN 0x30FF

jdgleaver commented:
Wonderful! You are a legend :)

Thank you so much for updating the code as well. As soon as I get my hands on a dev machine tomorrow, I will build and PR this.

trngaje commented:
> Thank you so much for updating the code as well. As soon as I get my hands on a dev machine tomorrow, I will build and PR this.

I'm just happy if you're satisfied.

jdgleaver commented:
It is a pleasure to work with you :)

jdgleaver commented:
I've made the PRs:

https://github.com/libretro/RetroArch/pull/11690
https://github.com/libretro/retroarch-assets/pull/395

Thank you again for this wonderful contribution! :)

trngaje commented:
@jdgleaver 
I have a question.
How can I delete my branch (ckj-bitmap-fonts)?
Where is the most effective place to ask programming questions about retroarch?
-Recently, I used uinput to insert the joypad input(udev) into the retroarch, but there is no response.
-Debug messages are being added and tracked but not easy.

trngaje commented:
> Thank you again for this wonderful contribution! :)

I'm just your hands and feet.

jdgleaver commented:
@trngaje 

> How can I delete my branch (ckj-bitmap-fonts)?

This is quite easy. First you should close this PR, then run the following commands from inside your RetroArch directory:

```
git checkout master
git branch -D ckj-bitmap-fonts
git push --delete origin ckj-bitmap-fonts
git pull upstream master
git push origin master
```

That will leave everything nice and clean.

> Where is the most effective place to ask programming questions about retroarch?

Do you use Discord? If so, please join us! You can use the `JOIN OUR DISCORD` button at the bottom of this page: https://www.retroarch.com/?page=discord

This is the most active RA community. There is a `#programming` channel where you can talk about general RA coding issues. We're quite friendly :)
